src/Chunk.ts

```diff
@@ -673,160 +688,219 @@ export default class Chunk {
+				this.usedModules.push(module);
+
+				if (namespace.needsNamespaceBlock) {
+					const rendered = namespace.renderBlock(renderOptions);
+					if (namespace.renderFirst()) hoistedSource += '\n' + rendered;
+					else magicString.addSource(new MagicString(rendered));
 				}
+			}
+		}
 
-				timeStart('render format', 3);
-
-				const getPath = this.createGetPath(options);
-
-				if (intro) intro += '\n\n';
-				if (outro) outro = `\n\n${outro}`;
-
-				magicString = finalise(
-					this,
-					magicString.trim(),
-					{
-						exportMode,
-						getPath,
-						indentString,
-						intro,
-						outro,
-						dynamicImport: !!renderOptions.importMechanism
-					},
-					options
-				);
+		// resolve external module paths
+		for (let external of this.dependencies) {
+			if (!(external instanceof ExternalModule)) continue;
+			external.setRenderPath(options, this.entryModule && this.entryModule.id);
+		}
 
-				timeEnd('render format', 3);
+		if (hoistedSource) magicString.prepend(hoistedSource + '\n\n');
 
-				if (banner) magicString.prepend(banner + '\n');
-				if (footer) magicString.append('\n' + footer);
+		this.renderedSource = magicString.trim();
 
-				const prevCode = magicString.toString();
-				const bundleSourcemapChain: RawSourceMap[] = [];
+		if (
+			this.getExportNames().length === 0 &&
+			this.getImportIds().length === 0 &&
+			this.renderedSource.isEmpty()
+		) {
+			this.graph.warn({
+				code: 'EMPTY_BUNDLE',
+				message: 'Generated an empty bundle'
+			});
+		}
 
-				return transformBundle(prevCode, this.graph.plugins, bundleSourcemapChain, options).then(
-					(code: string) => {
-						let map: SourceMap;
+		this.hasDynamicImport = !!renderOptions.importMechanism;
 
-						if (options.sourcemap) {
-							timeStart('sourcemap', 3);
+		this.renderedDeclarations = {
+			dependencies: this.getChunkDependencyDeclarations(options),
+			exports: this.getChunkExportDeclarations()
+		};
+	}
 
-							let file = options.file ? options.sourcemapFile || options.file : this.id;
-							if (file) file = resolve(typeof process !== 'undefined' ? process.cwd() : '', file);
+	generateNamePreserveModules(preserveModulesRelativeDir: string) {
+		return (this.id = normalize(relative(preserveModulesRelativeDir, this.entryModule.id)));
+	}
 
-							if (
-								this.graph.hasLoaders ||
-								this.graph.plugins.find(plugin =>
-									Boolean(plugin.transform || plugin.transformBundle)
-								)
-							) {
-								let decodedMap = magicString.generateDecodedMap({});
-								map = collapseSourcemaps(this, file, decodedMap, usedModules, bundleSourcemapChain);
-							} else {
-								map = magicString.generateMap({ file, includeContent: true });
-							}
+	generateName(pattern: string, addons: Addons, existingNames: { [name: string]: boolean }) {
+		let ext = extname(pattern);
+		pattern = pattern.substr(0, pattern.length - ext.length);
+		if (ext !== '.js' && ext !== '.mjs') {
+			pattern += ext;
+			ext = '.js';
+		}
```

Comment:
A reasonable approach but I would expect people to know what they are doing if they specify an extension. Extensions are not part of the JavaScript spec anyway. So I would prefer the following behaviour:
* No extension provided: Either add `.js` or - add no extension at all! Open for discussion here. Otherwise there is actually no way to create a chunk without an extension (with `/bin/rollup` we even create an extension-less file ourselves)
* An extension is provided: Do not add another extension as it quite possible this is intentional and there may be other extensions beyond (m)js that could make sense for JavaScript.